SignalServiceKit/src/Storage/OWSSessionStorage.m

```diff
@@ -0,0 +1,161 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSSessionStorage.h"
+#import "OWSFileSystem.h"
+#import "OWSStorage+Subclass.h"
+#import "TSDatabaseView.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString *const OWSSessionStorageExceptionName_CouldNotCreateDatabaseDirectory
+    = @"OWSSessionStorageExceptionName_CouldNotCreateDatabaseDirectory";
+
+#pragma mark -
+
+@interface OWSSessionStorage ()
+
+@property (nonatomic, readonly) YapDatabaseConnection *dbConnection;
+
+@property (atomic) BOOL areAsyncRegistrationsComplete;
+@property (atomic) BOOL areSyncRegistrationsComplete;
+
+@end
+
+#pragma mark -
+
+@implementation OWSSessionStorage
+
++ (instancetype)sharedManager
+{
+    static OWSSessionStorage *sharedManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedManager = [[self alloc] initStorage];
+
+#if TARGET_OS_IPHONE
+        [OWSSessionStorage protectFiles];
+#endif
+    });
+    return sharedManager;
+}
+
+- (instancetype)initStorage
+{
+    self = [super initStorage];
+
+    if (self) {
+        _dbConnection = self.newDatabaseConnection;
+
+        self.dbConnection.objectCacheEnabled = NO;
+#if DEBUG
+        self.dbConnection.permittedTransactions = YDB_AnySyncTransaction;
+#endif
+
+        OWSSingletonAssert();
+    }
+
+    return self;
+}
+
+- (StorageType)storageType
+{
+    return StorageType_Primary;
+}
+
+- (void)resetStorage
+{
+    _dbConnection = nil;
+
+    [super resetStorage];
+}
+
+- (void)runSyncRegistrations
+{
+    // Synchronously register extensions which are essential for views.
+    [TSDatabaseView registerCrossProcessNotifier:self];
+
+    OWSAssert(!self.areSyncRegistrationsComplete);
+    self.areSyncRegistrationsComplete = YES;
+}
+
+- (void)runAsyncRegistrationsWithCompletion:(void (^_Nonnull)(void))completion
+{
+    OWSAssert(completion);
+
+    // Asynchronously register other extensions.
+    //
+    // All sync registrations must be done before all async registrations,
+    // or the sync registrations will block on the async registrations.
+
+    // Block until all async registrations are complete.
+    [self.newDatabaseConnection asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
+        OWSAssert(!self.areAsyncRegistrationsComplete);
+
```

Comment:
So there just aren't any async registrations for the session store (yet)?

Reply:
Right. `OWSSessionStorage` so far just contains `SessionStore` and `IdentityStore` state.  I think it may eventually need to contain more.